File: src/main/java/com/blockout/competitions/models/CompetitionAssociation.java
package com.blockout.competitions.models;

import jakarta.persistence.*;
import lombok.*;
import java.time.LocalDateTime;

import com.blockout.competitions.models.enums.Category;

@Entity
@Table(name = "competition_association", uniqueConstraints = {
    @UniqueConstraint(columnNames = { "pool_id", "team_id", "category" }, name = "uix_pool_team_category")
})
@Data
@Builder(toBuilder = true)
@NoArgsConstructor
@AllArgsConstructor
public class CompetitionAssociation {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "pool_id", nullable = false)
    private Long poolId;

    @Column(name = "team_id", nullable = false)
    private Long teamId;

    @Column(name = "club_id", nullable = false)
    private String clubId;

    @Enumerated(EnumType.STRING)
    @Column(name = "category", nullable = false)
    private Category category;

    @Builder.Default
    @Column(name = "active", nullable = false)
    private Boolean active = true;

    @Builder.Default
    @Column(name = "points", nullable = false)
    private Integer points = 0;

    @Builder.Default
    @Column(name = "played", nullable = false)
    private Integer played = 0;

    @Builder.Default
    @Column(name = "wins", nullable = false)
    private Integer wins = 0;

    @Builder.Default
    @Column(name = "losses", nullable = false)
    private Integer losses = 0;

    @Builder.Default
    @Column(name = "wins_three_to_zero", nullable = false)
    private Integer winsThreeToZero = 0;

    @Builder.Default
    @Column(name = "wins_three_to_one", nullable = false)
    private Integer winsThreeToOne = 0;

    @Builder.Default
    @Column(name = "wins_three_to_two", nullable = false)
    private Integer winsThreeToTwo = 0;

    @Builder.Default
    @Column(name = "losses_zero_to_three", nullable = false)
    private Integer lossesZeroToThree = 0;

    @Builder.Default
    @Column(name = "losses_one_to_three", nullable = false)
    private Integer lossesOneToThree = 0;

    @Builder.Default
    @Column(name = "losses_two_to_three", nullable = false)
    private Integer lossesTwoToThree = 0;

    @Builder.Default
    @Column(name = "won_sets", nullable = false)
    private Integer wonSets = 0;

    @Builder.Default
    @Column(name = "lost_sets", nullable = false)
    private Integer lostSets = 0;

    @Builder.Default
    @Column(name = "won_points", nullable = false)
    private Integer wonPoints = 0;

    @Builder.Default
    @Column(name = "lost_points", nullable = false)
    private Integer lostPoints = 0;

    @Builder.Default
    @Column(name = "points_penalty", nullable = false)
    private Integer pointsPenalty = 0;

    @Builder.Default
    @Column(name = "coef_sets", nullable = false)
    private Double coefSets = 0.0;

    @Builder.Default
    @Column(name = "coef_points", nullable = false)
    private Double coefPoints = 0.0;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "last_update")
    private LocalDateTime lastUpdate;

    @PrePersist
    public void prePersist() {
        createdAt = LocalDateTime.now();
        lastUpdate = LocalDateTime.now();
    }

    @PreUpdate
    public void preUpdate() {
        lastUpdate = LocalDateTime.now();
    }
}

File: apps/backend/competition-scraper/scraper/infrastructure/blockout/competition_association.py
from dataclasses import dataclass
from datetime import datetime


@dataclass(frozen=True)
class BulkDeactivatePoolsInternalRequest:
    """Exact request accepted by competition-service's pool cleanup route."""

    missingPoolIds: list[int]


@dataclass(frozen=True)
class BulkDeactivateTeamsInternalRequest:
    """Exact request accepted by competition-service's team cleanup route."""

    missingTeamIds: list[int]


@dataclass
class CompetitionAssociationInternalResponse:
    """Complete competition-service association and ranking representation."""

    poolId: int
    teamId: int
    clubId: str
    id: int | None = None
    active: bool = True
    played: int = 0
    wins: int = 0
    losses: int = 0
    points: int = 0
    winsThreeToZero: int = 0
    winsThreeToOne: int = 0
    winsThreeToTwo: int = 0
    lossesZeroToThree: int = 0
    lossesOneToThree: int = 0
    lossesTwoToThree: int = 0
    wonSets: int = 0
    lostSets: int = 0
    wonPoints: int = 0
    lostPoints: int = 0
    pointsPenalty: int = 0
    coefSets: float = 0.0
    coefPoints: float = 0.0
    createdAt: datetime | None = None
    lastUpdate: datetime | None = None
